SSA/Projects/DCE/DCE.lean

```diff
@@ -0,0 +1,492 @@
+              simp[HCOM]
+              intros V
+              simp[ICom.denote]
+              apply hbody
+            ⟩⟩
+          let ⟨Γ'', hom'', ⟨com'', hcom''⟩⟩
+            :   Σ (Γ'' : Ctxt Ty) (hom: Ctxt.Hom Γ'' Γ'), { com'' : ICom Op Γ'' t //  ∀ (V' : Γ'.Valuation), com'.denote V' = com''.denote (V'.hom hom)} :=
+            dce_ com' -- recurse into `com'`, which contains *just* the `body`, not the `let`, and return this.
+          ⟨Γ'', hom''.composeRange hom', com'', by
+            intros V
+            rw[← HCOM]
+            rw[hcom']
+            rw[hcom'']
+            rfl⟩
+/-
+decreasing_by {
+  simp[invImage, InvImage, WellFoundedRelation.rel, Nat.lt_wfRel]
+  sorry -- Lean bug: *no goals to be solved*?!
+}
+-/
+
+/-- This is the real entrypoint to `dce` which unfolds the type of `dce_`, where we play the `DCEType` trick
+to convince Lean that the output type is in fact inhabited. -/
+def dce [OpSignature Op Ty] [OpDenote Op Ty]  {Γ : Ctxt Ty} {t : Ty} (com : ICom Op Γ t) :
+  Σ (Γ' : Ctxt Ty) (hom: Ctxt.Hom Γ' Γ),
+    { com' : ICom Op Γ' t //  ∀ (V : Γ.Valuation), com.denote V = com'.denote (V.hom hom)} :=
+  dce_ com
+
+/-- A version of DCE that returns an output program with the same context. It uses the context
+   morphism of `dce` to adapt the result of DCE to work with the original context -/
+def dce' [OpSignature Op Ty] [OpDenote Op Ty]  {Γ : Ctxt Ty} {t : Ty} (com : ICom Op Γ t) :
+    { com' : ICom Op Γ t //  ∀ (V : Γ.Valuation), com.denote V = com'.denote V} :=
+  let ⟨ Γ', hom, com', hcom'⟩ := dce_ com
+  ⟨com'.changeVars hom, by
+    intros V
+    rw[hcom']
+    simp[Ctxt.Valuation.hom]⟩
+
+namespace Examples
+
+/-- A very simple type universe. -/
+inductive ExTy
+  | nat
+  | bool
+  deriving DecidableEq, Repr
+
+@[reducible]
+instance : Goedel ExTy where
+  toType
+    | .nat => Nat
+    | .bool => Bool
+
+inductive ExOp :  Type
+  | add : ExOp
+  | beq : ExOp
+  | cst : ℕ → ExOp
+  deriving DecidableEq
+
+instance : OpSignature ExOp ExTy where
+  signature
+    | .add    => ⟨[.nat, .nat], [], .nat⟩
+    | .beq    => ⟨[.nat, .nat], [], .bool⟩
+    | .cst _  => ⟨[], [], .nat⟩
+
+@[reducible]
+instance : OpDenote ExOp ExTy where
+  denote
+    | .cst n, _, _ => n
+    | .add, .cons (a : Nat) (.cons b .nil), _ => a + b
+    | .beq, .cons (a : Nat) (.cons b .nil), _ => a == b
+
+def cst {Γ : Ctxt _} (n : ℕ) : IExpr ExOp Γ .nat  :=
+  IExpr.mk
+    (op := .cst n)
+    (ty_eq := rfl)
+    (args := .nil)
+    (regArgs := .nil)
+
+def add {Γ : Ctxt _} (e₁ e₂ : Ctxt.Var Γ .nat) : IExpr ExOp Γ .nat :=
+  IExpr.mk
+    (op := .add)
+    (ty_eq := rfl)
+    (args := .cons e₁ <| .cons e₂ .nil)
+    (regArgs := .nil)
+
+attribute [local simp] Ctxt.snoc
+
+def ex1_pre_dce : ICom ExOp ∅ .nat :=
+  ICom.lete (cst 1) <|
+  ICom.lete (cst 2) <|
+  ICom.ret ⟨0, by simp [Ctxt.snoc]⟩
+
+/-- TODO: how do we evaluate 'ex1_post_dce' within Lean? :D -/
+def ex1_post_dce : ICom ExOp ∅ .nat := (dce' ex1_pre_dce).val
```

Comment:
Can someone tell me how I actually evaluate this definition?

Reply:
Or do you mean, how do we actually reason about the result of `dce'`? That is a bit more difficult

Reply:
You can use `decreasing_by sorry` at the end instead of `partial def`, and then you will be able to unfold things.

Reply:
That doesn't work, see that there's a commented out block of code doing exactly this marked as `Lean bug`.